velox/type/Filter.cpp

```diff
@@ -458,6 +472,59 @@ bool MultiRange::testBytesRange(
   return false;
 }
 
+std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
+  switch (other->kind()) {
+    case FilterKind::kMultiRange: {
+      const MultiRange* multiRangeOther = static_cast<const MultiRange*>(other);
+      bool bothNullAllowed = nullAllowed_ && other->testNull();
+      bool bothNanAllowed = nanAllowed_ && multiRangeOther->nanAllowed_;
+      std::vector<std::unique_ptr<Filter>> merged;
+      for (auto const& filter : this->filters()) {
+        for (auto const& filterOther : multiRangeOther->filters()) {
+          auto innerMerged = filter->mergeWith(filterOther.get());
+          switch (innerMerged->kind()) {
+            case FilterKind::kAlwaysFalse:
+            case FilterKind::kIsNull:
+              break;
+            default:
+              merged.push_back(std::move(innerMerged));
+          }
+        }
+      }
+
+      if (merged.empty()) {
+        return nullOrFalse(bothNullAllowed);
+      } else if (merged.size() == 1) {
+        return merged[0]->clone(bothNullAllowed);
```

Comment:
Why not return std::move(merged.front())

Reply:
Yes. This is just ensure that the only filter matches the top level `nullAllowed`. For MultiRange merge, we will ignore inner filters' nullness.